engine/src/test/scala/cromwell/services/KeyValueServiceActorSpec.scala

```diff
@@ -0,0 +1,186 @@
+
+    val future = for {
+      _ <- dataAccess.createWorkflow(descriptor, Nil, descriptor.namespace.workflow.calls, localBackend)
+      r1 <- ask(kvActor, put).mapTo[KvPutSuccess]
+      _ = r1 shouldEqual KvPutSuccess(put)
+      r4 <- ask(kvActor, get).mapTo[KvPair]
+      _ = r4 shouldEqual KvPair(callAKey, "k", None)
+    } yield ()
+
+    Await.result(future, 2.seconds)
+  }
+
+  it should "partition keys by call" in {
+    val descriptor = makeWorkflowDescriptor(sources)
+    val kvActor = makeKeyValueActor(dataAccess, cromwellConfig, descriptor)
+    val callAKey = getBackendJobDescriptorKey(descriptor, "a")
+    val callBKey = getBackendJobDescriptorKey(descriptor, "b")
+    val putA = KvPut(callAKey, "k", Option("value_a"))
+    val putB = KvPut(callBKey, "k", Option("value_b"))
+    val getA = KvGet(callAKey, "k")
+    val getB = KvGet(callBKey, "k")
+
+    val future = for {
+      _ <- dataAccess.createWorkflow(descriptor, Nil, descriptor.namespace.workflow.calls, localBackend)
+      r0 <- ask(kvActor, putA).mapTo[KvPutSuccess]
+      _ = r0 shouldEqual KvPutSuccess(putA)
+      r1 <- ask(kvActor, getB).mapTo[KvKeyLookupFailed]
+      _ = r1.action shouldEqual getB
+      r2 <- ask(kvActor, putB).mapTo[KvPutSuccess]
+      _ = r2 shouldEqual KvPutSuccess(putB)
+      r3 <- ask(kvActor, getA).mapTo[KvPair]
+      _ = r3 shouldEqual KvPair(callAKey, "k", Option("value_a"))
+      r4 <- ask(kvActor, getB).mapTo[KvPair]
+      _ = r4 shouldEqual KvPair(callBKey, "k", Option("value_b"))
+    } yield ()
+
+    Await.result(future, 2.seconds)
+  }
+
+  it should "partition keys by workflow" in {
+    val descriptor0 = makeWorkflowDescriptor(sources)
+    val kv0 = makeKeyValueActor(dataAccess, cromwellConfig, descriptor0)
+    val descriptor1 = makeWorkflowDescriptor(sources)
+    val kv1 = makeKeyValueActor(dataAccess, cromwellConfig, descriptor1)
+
+    val callA0 = getBackendJobDescriptorKey(descriptor0, "a")
+    val callA1 = getBackendJobDescriptorKey(descriptor1, "a")
+    val putA0 = KvPut(callA0, "k", Option("v"))
+    val putA1 = KvPut(callA1, "k", Option("v"))
+    val getA0 = KvGet(callA0, "k")
+    val getA1 = KvGet(callA1, "k")
+
+    val future = for {
+      _ <- dataAccess.createWorkflow(descriptor0, Nil, descriptor0.namespace.workflow.calls, localBackend)
+      _ <- dataAccess.createWorkflow(descriptor1, Nil, descriptor1.namespace.workflow.calls, localBackend)
+      r0 <- ask(kv0, putA0).mapTo[KvPutSuccess]
+      _ = r0 shouldEqual KvPutSuccess(putA0)
+      r1 <- ask(kv1, getA1).mapTo[KvKeyLookupFailed]
+      _ = r1.action shouldEqual getA1
+      r2 <- ask(kv1, putA1).mapTo[KvPutSuccess]
+      _ = r2 shouldEqual KvPutSuccess(putA1)
+      r3 <- ask(kv0, getA0).mapTo[KvPair]
+      _ = r3 shouldEqual KvPair(callA0, "k", Option("v"))
+      r4 <- ask(kv1, getA1).mapTo[KvPair]
+      _ = r4 shouldEqual KvPair(callA1, "k", Option("v"))
+    } yield ()
+
+    Await.result(future, 2.seconds)
+  }
+}
```

Comment:
@geoffjentry will at some point comment that @kshakir will at some point comment about newlines here :)